File: tests/grpc/server_tls.nim
{.define: ssl.}

import ../../src/nimproto3
importProto3 currentSourcePath.parentDir & "/test_service.proto"

# Service handlers
proc handleSimpleTest(stream: GrpcStream) {.async.} =
  # 1. Read Request (Unary = Read 1 message)
  let msgOpt = await stream.recvMsg()

  if msgOpt.isNone:
    echo "[Service] Client closed without sending data?"
    return

  let input = msgOpt.get()
  let req = fromBinary(TestRequest, input)

  # Demonstrate reading Metadata (Headers)
  let auth = stream.headers.getOrDefault("authorization", "none")
  echo "[Service] Received: ", req.message, " (Counter: ", req.counter,
      ") | Auth: ", auth

  # 2. Logic
  let reply = TestReply(
    response: "Server says: " & req.message.toUpperAscii(),
    received: true
  )

  # 3. Send Response (Unary = Send 1 message)
  await stream.sendMsg(toBinary(reply))

  # When this async proc finishes, the server automatically closes the stream
  # and sends the Trailing Headers (Status: OK).

# Example of a Server Streaming handler (returning multiple items)
proc handleStreamTest(stream: GrpcStream) {.async.} =
  # Read incoming requests loop (Bidirectional or Client Stream)
  while true:
    let msgOpt = await stream.recvMsg()
    if msgOpt.isNone: break # End of Stream

    let req = fromBinary(TestRequest, msgOpt.get())
    echo "[Service] Stream item: ", req.message

    # Send a reply immediately (Echo)
    let reply = TestReply(response: "Echo: " & req.message, received: true)
    await stream.sendMsg(toBinary(reply))

# =============================================================================
# MAIN SERVER
# =============================================================================

when isMainModule:
  import std/asyncdispatch
  
  proc runServer() {.async.} =
    echo "================================================================================"
    echo "Nim gRPC Server with TLS - Local Test"
    echo "================================================================================"
    
    # Create server with TLS support
    let server = newGrpcServer(
        50051,
        CompressionIdentity,
        certFile = currentSourcePath.parentDir() / "server.crt",
        keyFile = currentSourcePath.parentDir() / "server.key"
    )

    server.registerHandler("/TestService/SimpleTest", handleSimpleTest)
    server.registerHandler("/TestService/StreamTest", handleStreamTest)
    
    echo "[Server] Starting TLS server on port 50051..."
    echo "[Server] Waiting for connections..."
    await server.serve()
  
  waitFor runServer()

File: tests/grpc/server_tls.py
#!/usr/bin/env python3
"""
gRPC Server with TLS for testing Nim gRPC client
"""
import grpc
from concurrent import futures
import sys
import os

# Import generated protobuf files
import test_service_pb2
import test_service_pb2_grpc

class TestServiceServicer(test_service_pb2_grpc.TestServiceServicer):
    def SimpleTest(self, request, context):
        print(f"[Server] Received SimpleTest: message='{request.message}', counter={request.counter}")
        return test_service_pb2.TestReply(
            response=f"Echo: {request.message}",
            received=True
        )
    
    def StreamTest(self, request_iterator, context):
        print(f"[Server] Received StreamTest (bidirectional)")
        for request in request_iterator:
            print(f"[Server] Got message: '{request.message}', counter={request.counter}")
            yield test_service_pb2.TestReply(
                response=f"Echo: {request.message}",
                received=True
            )

def serve():
    # Read certificate and key
    cert_dir = os.path.dirname(os.path.abspath(__file__))
    with open(os.path.join(cert_dir, 'server.crt'), 'rb') as f:
        cert = f.read()
    with open(os.path.join(cert_dir, 'server.key'), 'rb') as f:
        key = f.read()
    
    # Create server credentials
    server_credentials = grpc.ssl_server_credentials([(key, cert)])
    
    # Create server
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    test_service_pb2_grpc.add_TestServiceServicer_to_server(TestServiceServicer(), server)
    
    # Start server with TLS
    port = 50051
    server.add_secure_port(f'[::]:{port}', server_credentials)
    server.start()
    
    print(f"[Server] gRPC Server with TLS started on port {port}")
    print(f"[Server] Using cert: server.crt, key: server.key")
    
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print("\n[Server] Shutting down...")
        server.stop(0)

if __name__ == '__main__':
    serve()
